Sources/MapboxNavigation/OverlayController.swift

```diff
@@ -0,0 +1,390 @@
+                
+                for feature in features {
+                    var lineStrings: [LineString] = []
+                    
+                    if let line = feature.geometry.value as? LineString {
+                        lineStrings.append(line)
+                    } else if let multiLine = feature.geometry.value as? MultiLineString {
+                        for coordinates in multiLine.coordinates {
+                            lineStrings.append(LineString(coordinates))
+                        }
+                    }
+                    
+                    for lineString in lineStrings {
+                        let lookAheadDistance: CLLocationDistance = 10
+                        guard let pointAheadFeature = lineString.sliced(from: closestCoordinate)!.coordinateFromStart(distance: lookAheadDistance) else { continue }
+                        guard let pointAheadUser = slicedLine.coordinateFromStart(distance: lookAheadDistance) else { continue }
+                        guard let reversedPoint = LineString(lineString.coordinates.reversed()).sliced(from: closestCoordinate)!.coordinateFromStart(distance: lookAheadDistance) else { continue }
+                        
+                        let distanceBetweenPointsAhead = pointAheadFeature.distance(to: pointAheadUser)
+                        let distanceBetweenReversedPoint = reversedPoint.distance(to: pointAheadUser)
+                        let minDistanceBetweenPoints = min(distanceBetweenPointsAhead, distanceBetweenReversedPoint)
+                        
+                        if minDistanceBetweenPoints < smallestLabelDistance {
+                            smallestLabelDistance = minDistanceBetweenPoints
+                            
+                            let roadNameRecord = self.roadFeature(for: feature)
+                            currentShieldName = roadNameRecord.shieldName
+                            currentName = roadNameRecord.roadName
+                        }
+                    }
+                }
+                
+                let hasWayName = currentName != nil || currentShieldName != nil
+                if smallestLabelDistance < 5 && hasWayName {
+                    if let currentShieldName = currentShieldName {
+                        self.navigationView.wayNameView.attributedText = currentShieldName
+                    } else if let currentName = currentName {
+                        self.navigationView.wayNameView.text = currentName
+                    }
+                    self.navigationView.wayNameView.isHidden = false
+                } else {
+                    self.navigationView.wayNameView.isHidden = true
+                }
+            case .failure:
+                NSLog("Failed to find visible features.")
+            }
+        })
+    }
+    
+    private func roadFeature(for line: Feature) -> (roadName: String?, shieldName: NSAttributedString?) {
```

Comment:
This attributed string building logic is specific to the layout needed by WayNameView, so it should live in that class.

Reply:
Done